File: src/main/java/komu/blunt/types/TypeGen.java
package komu.blunt.types;

import java.util.List;
import java.util.Set;

public final class TypeGen extends Type {
    
    private final int index;

    public TypeGen(int index) {
        this.index = index;
    }

    @Override
    public TypeGen apply(Substitution substitution) {
        return this;
    }

    @Override
    protected Type instantiate(List<TypeVariable> vars) {
        return vars.get(index);
    }

    @Override
    public void addTypeVariables(Set<TypeVariable> result) {
        // no variables here
    }

    @Override
    protected Kind getKind() {
        throw new RuntimeException("can't access kind of TypeGen");
    }

    @Override
    public boolean hnf() {
        throw new RuntimeException("should not call hnf for TypeGen");
    }

    @Override
    protected String toString(final int precedence) {
        return "TypeGen[" + index + "]";
    }
}

File: src/main/kotlin/komu/blunt/types/TypeGen.kt
package komu.blunt.types

import komu.blunt.types.checker.Substitution

class TypeGen(private val index: Int) : Type() {

    override fun apply(substitution: Substitution) = this
    override fun instantiate(vars: List<TypeVariable>) = vars[index]
    override fun addTypeVariables(result: MutableSet<TypeVariable>) { }
    override val kind: Kind
        get() = throw RuntimeException("can't access kind of TypeGen")
    override fun hnf(): Boolean = throw RuntimeException("should not call hnf for TypeGen")
    override fun toString(precedence: Int) = "TypeGen[$index]"
}

